### PR TITLE
Working on Razor pages 2.0

### PR DESCRIPTION
**Pull request by jtkech**
**Where i am**.

- So, for testing we have a razor page in `Orchard.Demo/Pages/Hello.cshtml`.

- And a `_ViewImports` for taghelpers and to inherit of a custom `RazorPages.Page`, see `Page.cs`.

- We can access it through this url `/Orchard.Demo/Pages/Hello`.

- If `Orchard.Demo` is not enabled, we get a `NotFound` page, see `ModularRazorPageFilter`.

- The theme layout is naturally used because of our precompiled views (as `ThemeViewStart`).

- The `Hello.cshtml` page defines a controller and a simple model which is used when rendering.

- We can use taghelpers, for testing our `zone` taghelper is used to put an image in the header zone.

- We can render shapes, for testing our `DateTimeShape` is used to render the current date.

**Todo**.

- See if there is a way to map a page route, e.g to a simpler one. Maybe not intended to be done.

- In a page we can render other `partials views`, see if we can override them through a theme. Note: I think we can't override the page itself.

- To override `partial views`, i think we need to update our theme view location expander. In the context of a `RazorPages.Page`, `context.AreaName` is null but we could use `context.PageName`.

### CODE REVIEW

File: src/OrchardCore/Orchard.Mvc.Core/ModularRazorPageFilter.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        private readonly IHttpContextAccessor _httpContextAccessor;
+
+        public ModularRazorPageFilter(IHttpContextAccessor httpContextAccessor)
+        {
+            _httpContextAccessor = httpContextAccessor;
+        }
+
+        public void OnPageHandlerExecuted(PageHandlerExecutedContext context)
+        {
+            var shellBluePrint = _httpContextAccessor.HttpContext.RequestServices.GetRequiredService<ShellBlueprint>();
```
Comment (sebastienros) [inline src/OrchardCore/Orchard.Mvc.Core/ModularRazorPageFilter.cs]: Looks hacky.

**Comment by sebastienros** [comment]
See this issue for more ideas on how to do things https://github.com/aspnet/Mvc/issues/6671#issuecomment-323133037

Also there is a MapPageRoute that might be useful.

File: src/OrchardCore/Orchard.Mvc.Core/ModularRazorPageFilter.cs (hunk 1 above)
Comment (jtkech) [reply src/OrchardCore/Orchard.Mvc.Core/ModularRazorPageFilter.cs]: Yes, quickly done for testing, but look at `ShellFeatureApplicationPart` it uses exactly the same line of code.
I will commit a version which inject directly `IShellFeaturesManager` to get enabled features.

**Comment by jtkech** [comment]
Ok, i will take a look, thanks.

For infos, i'm looking about watching. Because we want pages in different modules, right now for testing i specify `/Packages` as the razor pages root directory. Then razor search pages with the pattern `/Packages/**/*.cshtml`. The result is that razor watches again all razor files. Not so good.

Moreover it requests directory contents, it was not the case with only regular razor files, and it also requests change tokens with wildcard characters. So, maybe we have to adapt our custom file provider to make our watching working. Or find another way, there are many other things i didn't see yet.

Note: Now, when i update a page file (or any other razor file) directly under the `Packages` folder, i don't know why but it calls our `ShellFeatureApplicationPart.Types` which uses `_httpContextAccessor.HttpContext.RequestServices`. And unfortunately, at this point it was null, so i added a null check.

**Comment by jtkech** [comment]
- So, we use `ModularPageViewEnginePathFilter` which is an `IAsyncPageFilter`, this to filter pages according to the enabled modules.

- Now, the above filter is added through `ModularPageApplicationModelProvider` which is an `IPageApplicationModelProvider` and where we could add other filters in the future.

- Then, we use `ModularPageRouteModelConvention` which is an `IPageRouteModelConvention`, this to use e.g `/Orchard.Demo/Hello` in place of `/Orchard.Demo/Pages/Hello`.

- Todo: I will try to find a way to prevent razorPages to watch all `.cshtml` files.

**Comment by jtkech** [comment]
@sebastienros 

- So, now the default convention is to use the url `Orchard.Demo/Hello`.

- The demo module uses a `MapPageRoute()` to be able to just use `/hello`.

- Now, pages are also watched in modules projects while in dev.

- I needed to use a file watcher for each module project having a `/Pages` folder. But i think it's ok because it's only while in dev, only for modules referenced as projects, and having a `/Pages` folder ...

**Comment by jtkech** [comment]
Todo: be able to override a razor page through the theme using a `PageViewLocationExpander`.

**Comment by jtkech** [comment]
@sebastienros 

- In fact, the default `PageViewLocationExpander` is not intended to expand the locations of a page itself, but to **expand the locations of a view which is rendered through a given page**. This by starting from the page location and doing an ascending directory search.

- So, the view can be in the page folder but the problem is that they will both use the same `_ViewImports` so the view will also inherit of our custom `RazorPages.Page`. A solution is to specify `Razor.RazorPage` directly in the view, or, as here, **to use a separate folder** with its own imports file.

- So, here the suggestion (can be changed) is to specify to the ascending search an additional `PageViews` folder. This folder can be under the `Pages` folder e.g `Orchard.Demo/Pages/PageViews`, or at the root of the module (because of the ascending search) e.g `Orchard.Demo/PageViews`.

- Then, the view can be overriden by using e.g this folder `TheBlogTheme/PageViews/Orchard.Demo`.

**Note**: We could use `Views` in place of `PageViews` but because of the ascending search, the view might be found in the regular `Module/Views` folder. Maybe good but here it is to separate things.

**Note**: If a theme defines its own page which renders a view in `TheTheme/PageViews/MyView.cshtml`, this view might be used by a page defined in another module. A solution is to use here another name than `PageViews`.

But then, maybe too many names: `/Pages` for pages, `Views` for regular views, `PageViews` for views rendered through a page, and then another name than `PageViews` when in a theme and to override view used by pages defined in other modules, not defined in the theme itself.

Need more thinking.

**Comment by jtkech** [comment]
Here what i've opted to do.

- So, i was talking about pages which renders **partial views** by their names. I called them `PageViews` because when in the context of a page, there is a default `PageViewLocationExpander` used to find them.

- But to have a different `_ViewImports`, we need a different folder and expand the locations to find them. **I opted to put them in the `Views` folder** (but still in specific subfolders, see below) so they use the same `_ViewImports` as other views.

- We can't override the page itself in a theme but we can expand the locations to be able to override the partial views it uses. **Here also i opted to put them in the `Views` folder** (but still in specific subfolders).

- **Todo**: When overriding a partial view **i will inhibit part of the ascending searching**. This to prevent a partial view used by a page defined in the theme itself, to be used by a page of another module.

- Inside a module (could be a theme).

      - Pages
          - _ViewImports.cshtml
          - APage.cshtml
          - SubFolder
              - AnotherPage.cshtml // use the above imports
      - Views
          - PageViews
              - PartialView.cshtml // can be found by all pages of this module and uses the below imports
      - _ViewImports.cshtml
      - OtherView.cshtml

- Inside a theme

      - Views
          - Orchard.ModuleName
              -PageViews
                  - PartialView.cshtml // can be found by all pages of ModuleName and uses the below imports
      - _ViewImports.cshtml
      - OtherView.cshtml

**Pull request merged**